internal/server/middleware.go

```diff
@@ -81,42 +67,61 @@ func handleInfraDestinationHeader(c *gin.Context) error {
 	}
 }
 
-// TODO: remove duplicate in access package
-func getDB(c *gin.Context) *gorm.DB {
-	db, ok := c.MustGet("db").(*gorm.DB)
-	if !ok {
-		return nil
-	}
-	return db
-}
-
 // authenticatedMiddleware is applied to all routes that require authentication.
 // It validates the access key, and updates the lastSeenAt of the user, and
 // possibly also of the destination.
-func authenticatedMiddleware() gin.HandlerFunc {
+func authenticatedMiddleware(db *gorm.DB) gin.HandlerFunc {
 	return func(c *gin.Context) {
-		db := getDB(c)
-		authned, err := requireAccessKey(db, c.Request)
-		if err != nil {
-			sendAPIError(c, err)
-			return
-		}
+		withDBTxn(c.Request.Context(), db, func(tx *gorm.DB) {
```

Comment:
It's unfortunate this adds another level of indentation.

In the future, when we remove gorm, we'll have to replace this with calls to `Begin`, `Commit`, and `Rollback`. I expect that to change this code a little bit, but probably not too much.